### PR TITLE
Removes GenomeLoc references from ExomeReadCounts tool code and related ...

### PR DESCRIPTION
**Pull request by vruano**
...components:

Resolves issue #421.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java [hunk 1]
```diff
@@ -1,30 +1,60 @@
...
+    private final List<SimpleInterval> sortedIntervals;
+
+    /**
+     * Compares intervals returning 0 only if they are exactly the same.
+     *
+     * <p>
+     *    Given intervals {@code a} and {@code b} on the same contig, this comparator will return:
+     *     <ul>
+     *         <li>-1 if {@code a}'s start is before {@code b}'s and if the same, {@code a}'s end is before {@code b}'s,</li>
+     *         <li>1 if {@code b}'s start is before {@code a}'s and if the same, {@code b}'s end is before {@code a}'s or</li>
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: This description is really confusing.   It sounds like the algorithm is
```
if (a.start < b.start && a.end < b.end) )
    return -1
else if (b.start < a.start && b.end < a.end)
    return 1
else
   return 0
```

Which is not the what the comparator is doing.  

It would be clearer to say something like "First compare the start position of both intervals, if they're different compare then end position" or just literally copy the code from the actual comparator.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java [hunk 2]
```diff
@@ -216,16 +221,17 @@ protected CountColumns countColumns(final ExomeReadCounts tool) {
...
-
-    /**
      * Count matrix indexed by count column and then exon.
      */
     private int[][] counts;
 
     @Override
+    public ReadFilter makeReadFilter() {
+        return super.makeReadFilter()
+                .and(ReadFilterLibrary.MAPPED);
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java]: Seems like a good solution.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java [hunk 3]
```diff
@@ -216,16 +221,17 @@ protected CountColumns countColumns(final ExomeReadCounts tool) {
...
+    }
+
+    @Override
     public void onTraversalStart() {
         super.onTraversalStart();
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java]: no need for the super call

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java [hunk 4]
```diff
@@ -238,13 +244,13 @@ public void onTraversalStart() {
...
         sampleCollection = new SampleCollection(getHeaderForReads());
 
         logger.log(Level.INFO, "Reading exons locations from intervals...");
-        final List<GenomeLoc> locations = IntervalUtils.genomeLocsFromLocatables(genomeLocFactory,
-                intervalArgumentCollection.getIntervals(sequenceDictionary));
-        exonCollection = new IntervalBackedExonCollection(locations);
+
+        exonCollection = new IntervalBackedExonCollection(
+                intervalArgumentCollection.getIntervals(sequenceDictionary),
+                (contig1,contig2) -> Integer.compare(sequenceDictionary.getSequenceIndex(contig1),sequenceDictionary.getSequenceIndex(contig2)));
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java]: This is clever, but why are you specifying the comparator here?  Is there ever a case when you would want a different comparator?

Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java]: Also, does the order of sorting actually matter?  Or is the fact that their is some ordering the important thing?

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java [hunk 5]
```diff
@@ -289,7 +295,8 @@ private PrintWriter openOutputWriter(final File output, final String headerText)
 
     @Override
     public void apply(final SAMRecord read, final ReferenceContext referenceContext, final FeatureContext featureContext) {
-        final GenomeLoc readLocation = genomeLocFactory.createGenomeLoc(read);
+
+        final SimpleInterval readLocation = new SimpleInterval(read.getContig(),read.getStart(),read.getEnd());
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java]: You shouldn't have to do this.  Could you add this to `SimpleInterval` as a copy constructor from `Locatable`

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java [hunk 6]
```diff
@@ -33,46 +63,114 @@
...
+     */
+    private Comparator<SimpleInterval> buildIntervalOverlapComparator(final Comparator<String> contigComparator) {
+
+        // It turns out that figuring out if there is an overlap is simply whether
+        // the cross-comparisons of start and end positions of both intervals is contradictory:
+        return Comparator.comparing(SimpleInterval::getContig, contigComparator)
+                .thenComparing((a, b) -> {
+                    final int startEndCompare = Integer.compare(a.getStart(),b.getEnd());
+                    final int endStartCompare = Integer.compare(a.getEnd(),b.getStart());
+                    return startEndCompare == endStartCompare ? startEndCompare : 0;
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: Would this be clearer?  
`` a.overlaps(b) : 0 ? Integer.compare(a.getStart(), b.getStart()``

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java [hunk 7]
```diff
@@ -33,46 +63,114 @@
...
+     *     intervals on different contigs.
+     * </p>
+     *
+     * @param intervals the input interval list. Is assumed to be sorted
+     * @param contigComparator comparator used to establish the absolute order
+     *                         between intervals in different contigs.
+     * @throws IllegalArgumentException if {@code intervals} is {@code null}.
+     */
+    public IntervalBackedExonCollection(final List<SimpleInterval> intervals,
+                                        final Comparator<String> contigComparator) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: Does the  contig ordering matter as long as the contigs are separated?

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java [hunk 8]
```diff
@@ -33,46 +63,114 @@
...
+        this.intervalsByName = composeIntervalsByName(sortedIntervals);
     }
 
     /**
-     * Throws a exception if there some interval location incoherence.
+     * Fails with an exception if the intervals collection has overlapping intervals.
+     * @param sortedIntervals
+     * @param overlapComparator
+     */
+    private static void checkForOverlaps(final List<SimpleInterval> sortedIntervals,
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: Why does this use the overlap comparator instead of just calling `SimpleInterval.overlaps`?

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java [hunk 9]
```diff
@@ -1,30 +1,60 @@
...
+    /**
+     * Compares intervals returning 0 if these overlap.
+     *
+     * <p>
+     *     Given intervals {@code a} and {@code b} on the same contig, this comparator will return:
+     *     <ul>
+     *         <li>0 if {@code a} and {@code b} overlap,</li>
+     *         <li>-1 if {@code a} is strictly before {@code b} and</li>
+     *         <li>1 if {@code b} is strictly before {@code b}</li>
+     *     </ul>
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: Add this to the javadoc here:

Note: this comparator imposes orderings that are inconsistent with equals.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java [hunk 10]
```diff
@@ -33,46 +63,114 @@
...
+            throw new IllegalArgumentException(
+                    String.format("input intervals contain at least two overlapping intervals: %s and %s",
+                            sortedIntervals.get(index-1),sortedIntervals.get(index)));
+        }
+    }
+
+    /**
+     * Given a contig ({@link String}) comparator it will compose interval comparator that will
+     * return 0 when both intervals are overlapping.
      *
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: add to the javadoc
 Note: this comparator imposes orderings that are inconsistent with equals."

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java [hunk 11]
```diff
@@ -368,11 +410,11 @@ private int extendSearchIndexForward(final GenomeLoc location, final int startIn
...
      */
-    private int extendSearchIndexBackwards(final GenomeLoc location, final int startIndex) {
-        final ListIterator<GenomeLoc> it = sortedIntervals.listIterator(startIndex);
+    private int extendSearchIndexBackwards(final SimpleInterval location, final int startIndex) {
+        final ListIterator<SimpleInterval> it = sortedIntervals.listIterator(startIndex);
         while (it.hasPrevious()) {
-            final GenomeLoc previous = it.previous();
-            if (previous.isBefore(location)) {
+            final SimpleInterval previous = it.previous();
+            if (intervalOverlapComparator.compare(previous,location) < 0) {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: I think it would be clear to just have `previous.getEnd() < previous.getStart()`

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java (hunk 9 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: Actually, after thinking about this a bit more, I don't think this is a valid comparator.

from the [comparator definition](https://docs.oracle.com/javase/7/docs/api/java/util/Comparator.html)

>implementor must ensure that compare(x, y)==0 implies that sgn(compare(x, z))==sgn(compare(y, z)) for all z.

this isn't true for this comparator, 
ex:
```
x = [1,3]
y = [2,6]
z = [5,8]

compare(x,y) == 0 
sgn(compare(x,z)) = -1
sgn(compare(y,z)) = 0
```

**Comment by lbergelson** [comment]
@vruano Is the sorting order of the chromosomes important?  Or is it only important that they are sorted in a searchable order?

**Comment by vruano** [comment]
The second. Will remove the order by sequenceDictionary.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java (hunk 11 above)
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: This may fail when previous is on a different chromosome (the previous one).

**Comment by vruano** [comment]
The sorting is as important as the user expectancy on the output order. In this particular case so far the only user is going to be ourselves (e.g. ReCapSeg). So I would say that is not that important. Will remove it (use the lexicographical order) in the reviewed code.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java (hunk 5 above)
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java]: I think this should be addressed in a different pull request as this one would not be the only place that  should be updated to use that new constructor.

**Comment by vruano** [comment]
I've just updated the code based on your comments so far. 

Please @lbergelson review, V.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java (hunk 11 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: good point, I didn't think that through properly

**Comment by vruano** [comment]
@lbergelson are you still reviewing this?

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java (hunk 5 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java]: It exists now!  Lets use it.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ExonCollection.java [hunk 12]
```diff
@@ -42,6 +43,24 @@
     E exon(final int index);
 
     /**
+     * Returns the name of an exon.
+     *
+     * <p>
+     *     This method is guaranteed to return the name of the exon in this collection as long as it is indeed
+     *     included in it.
+     * </p>
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ExonCollection.java]: These </p> look gross in the code.  Do they improve the javadoc substantially?

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java [hunk 13]
```diff
@@ -33,43 +35,65 @@
...
-    public IntervalBackedExonCollection(final List<GenomeLoc> intervals) {
+    public IntervalBackedExonCollection(final List<SimpleInterval> intervals) {
         if (intervals == null) {
             throw new IllegalArgumentException("the input intervals cannot be null");
         }
-        checkInputIntervalCoherence(intervals);
-        this.sortedIntervals = Collections.unmodifiableList(new ArrayList<>(intervals));
-        this.intervalsByName = composeIntervalsByName(intervals);
+        if (intervals.contains(null)) {
+            throw new IllegalArgumentException("the input cannot contain null");
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: Is this a real problem you ran into?

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java [hunk 14]
```diff
@@ -33,43 +35,65 @@
...
 
+
+    /**
+     * Comparator use to sort and search intervals.
+     * <p>
+     * Since we are guaranteed not to have overlaps between intervals (checked at construction),
+     * we only need to pay attention to the contig (sorted alphanumerically) and the start position.
+     * </p>
+     */
+    private Comparator<SimpleInterval> INTERVAL_COMPARATOR = (a,b) -> {
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: This could be written more succinctly as:

```
Comparator.comparing(SimpleInterval::getContig).thenComparingInt(SimpleInterval::getStart);
```

File: src/test/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollectionUnitTest.java [hunk 15]
```diff
@@ -79,11 +77,12 @@ public void setUp() {
                     break;
                 }
                 nonOverlappingExomeIntervals.add(
-                        ExomeToolsTestUtils.GENOME_LOC_FACTORY.createGenomeLoc(
+                        ExomeToolsTestUtils.createInterval(
                                 sequence.getSequenceName(), start, stop));
                 current = stop + 1;
             }
         }
+        Collections.sort(nonOverlappingExomeIntervals,((Comparator<SimpleInterval>)(a,b) -> a.getContig().compareTo(b.getContig())).thenComparing((a,b) -> Integer.compare(a.getStart(),b.getStart())));
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollectionUnitTest.java]: use the predefined comparator here too?

File: src/test/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollectionUnitTest.java [hunk 16]
```diff
@@ -91,6 +90,59 @@ public void testCorrectInitialization() {
         new IntervalBackedExonCollection(nonOverlappingExomeIntervals);
     }
 
+    @Test
+    public void testCorrectInitializationSortingContigsByName() {
+        final List<SimpleInterval> sortedLexicographically = nonOverlappingExomeIntervals.stream()
+                .sorted(((Comparator<SimpleInterval>) (a, b) -> a.getContig().compareTo(b.getContig())).thenComparing(
+                        (a, b) -> {
+                            final int startCompare = Integer.compare(a.getStart(), b.getStart());
+                            return startCompare != 0 ? startCompare : Integer.compare(a.getEnd(), b.getEnd());
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollectionUnitTest.java]: again, I think you can use the same comparator

File: src/test/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollectionUnitTest.java [hunk 17]
```diff
@@ -108,49 +160,26 @@ public void testIncorrectInitialization1() {
...
         new IntervalBackedExonCollection(intervals);
     }
 
     @Test(expectedExceptions = {IllegalArgumentException.class})
     public void testIncorrectInitialization6() {
-        final List<GenomeLoc> intervals = new ArrayList<>(nonOverlappingExomeIntervals);
-        intervals.add(1, intervals.get(0).getStopLocation());
+        final List<SimpleInterval> intervals = new ArrayList<>(nonOverlappingExomeIntervals);
+        final SimpleInterval interval = intervals.get(0);
+        intervals.add(1, new SimpleInterval(interval.getContig(),interval.getEnd(),interval.getEnd()));
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollectionUnitTest.java]: this can use the new constructor too

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ExonCollection.java (hunk 12 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ExonCollection.java]: On further reflection.  Leave the javadoc however you like it.  We should let @vdauwera decide what the right formatting of this sort of thing is.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java (hunk 13 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: Leave the check if you think it's useful.  I think it's unlikely to come up often, but it would be confusing when it happened.

File: src/test/java/org/broadinstitute/hellbender/tools/exome/ExomeToolsTestUtils.java [hunk 18]
```diff
@@ -34,42 +34,57 @@ protected static File getTestDataDir(){
      */
     protected final static SAMSequenceDictionary REFERENCE_DICTIONARY = SAMSequenceDictionaryExtractor.extractDictionary(REFERENCE_FILE);
 
+    protected static Comparator<SimpleInterval> createIntervalComparator() {
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/ExomeToolsTestUtils.java]: This is the second time a comparator like this shows up.  I think it should probably just be extracted out of here and put into `SimpleInterval`.

I think it can be more clearly expressed as 

```
 Comparator<SimpleInterval> comparator = Comparator.comparing(SimpleInterval::getContig)
.thenComparingInt(SimpleInterval::getStart)
.thenComparingInt(SimpleInterval::getEnd);
```

**Comment by lbergelson** [comment]
@vruano This is so much clearer now.  Much better to write your own search function that does what you want than to use the existing one with an invalid comparator and rely on the implementation to stay the same.  

The only issues left I think are to create a comparator in SimpleInterval and just use that, instead of having a lot of them scattered all over the place.  I think they're all compatible with doing the same thing sorting first by contig, then start, then end.   That and using the new SimpleInterval(Locatable) constructor when appropriate.

File: src/test/java/org/broadinstitute/hellbender/tools/exome/ExomeToolsTestUtils.java (hunk 18 above)
Comment (vruano) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/ExomeToolsTestUtils.java]: I was wondering why SimpleInterval is not  a Comparable<SimpleInterval>  and I would need to define this comparator at all. @droazen any comments?

Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/ExomeToolsTestUtils.java]: We didn't make it comparable because there isn't an inherent ordering on them.  I think it would be fine to provide a comparator though.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java (hunk 13 above)
Comment (vruano) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/IntervalBackedExonCollection.java]: This is necessary to avoid NPE propagation in case of a bug up-stream. Bugs are always going to be unlikely (or at least one hopes so), but if you remove this check you may as well remove any other IAE throwing conditional in the code for that matter.

File: src/test/java/org/broadinstitute/hellbender/tools/exome/ExomeToolsTestUtils.java (hunk 18 above)
Comment (vruano) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/ExomeToolsTestUtils.java]: So are you suggesting to create a "standard" comparator as a constant in SimpleInterval? Say
SimpleInterval.LEXICALLY_SORTED_CONTIG_COMPARATOR? A better suggestion for the name?

Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/ExomeToolsTestUtils.java]: Yeah, that's what I was suggesting.  I don't know what a better name would be, although that one is a monstrosity.

Comment (vruano) [inline src/test/java/org/broadinstitute/hellbender/tools/exome/ExomeToolsTestUtils.java]: Will call it LEXICOGRAPHICAL_ORDER_COMPARATOR

**Comment by vruano** [comment]
Addressed comments. Back to @lbergelson

**Comment by lbergelson** [comment]
Did you hang on to those comparator declarations in the test for a reason or were they just forgotten?  Other than that :+1: .  Rebase and merge when tests pass.  Don't wait for coveralls, it seems to be down today.

**Comment by lbergelson** [comment]
and coveralls is back
